app/controllers/application_controller.rb

```diff
@@ -60,6 +60,8 @@ def status
                  status: :not_acceptable
   end
 
+  rescue_from ActionController::InvalidAuthenticityToken, with: :handle_forgery_protection
```

Comment:
Why do we need this? Aren't all `InvalidAuthenticityToken` exceptions already handled by `protect_from_forgery`?

Reply:
I'm not super found of this approach. I think it can be confusing to see an error about CSRF token instead of seeing that auth was wrong in the logs in case we try to understand why some request failed. I can easily imagine debugging long time why CSRF failed when issue was actually wrong token.

Reply:
removed previous comment, I'm very confused about this. So we had the forgery protection custom strategy invoked already, which raised `ActionController::InvalidAuthenticityToken` but then we catch it to handle it. Doesn't it make sense for the forgery protection custom strategy to do that (i.e. return 403)? Or if not possible to do with a class, then handle with a method instead?

Reply:
@akostadinov Rails forgery protection can raise an exception but it doesn't handle it.

What I did is to add the old behavior for the login controller to all controllers (https://github.com/3scale/porta/pull/3543/files#diff-60b5cff5242a76414f7d7980e7a101164a369230dd8b96abdbfa3f10198894f8L15-L17)

So before:
|Login controller|All other controllers|
|--|--|
|`:exception`|`:reset`|

And now:
|Login controller|All other controllers|
|--|--|
|`:reset` + `:exception`|`:reset` + `:exception`|

The way to handle the exception is the same we were using before.

Reply:
I yet need to test how it looks like from the user perspective, but I am thinking - would it make sense to redirect to login page whenever `ActionController::InvalidAuthenticityToken` is caught? (rather than throwing a 403)

Just thinking out loud.

Reply:
@mayorova In fact, redirecting to login was the old behavior for the `:reset` method, (check https://github.com/3scale/porta/pull/3543/files#diff-ffe91beb10984c70928bc271d33d3235b27830610b76c17f7ad2449c246d0ee9L60). Since the session was reset, the redirect to the login screen was automatic. Not for the login screen in the developer portal though, since that one launched the exception, it returned a 403.

I think it's better to keep the 403 because the exception should never happen for legitimate users, only those being victims or request forgery, which should be very few. And for those cases, a clear Forbidden error gives more information about something going wrong than just redirecting to login without any error message. Also, since our protection method also resets the session, they will be redirected to login in their next request anyway.

Reply:
> I think it's better to keep the 403 because the exception should never happen for legitimate users, only those being victims or request forgery, which should be very few.

OK, yes, fair enough

Reply:
haven't tried but I think we can have `protect_from_forgery with: :method_name` so within a method we can both - check for token and handle it. Will avoid jumping through hoops. First raise and then handle is somehow disconnected way...But if can't be done clean, I can live with this.

Reply:
@akostadinov `:method_name` in that context means a class that implements `handle_unverified_request` (See the [docs](https://api.rubyonrails.org/classes/ActionController/RequestForgeryProtection/ClassMethods.html#method-i-protect_from_forgery)), which is what we do in this PR ([diff](https://github.com/3scale/porta/pull/3543/files#diff-3ed895612f90a4e34bd58d13370909cd3a508c91a08b02dcfef6e7fb22689835)). So we could force a redirect to login from `ExceptionAndResetStrategy.handle_unverified_request` but as I mentioned in my previous comment I don't think it's worth it.